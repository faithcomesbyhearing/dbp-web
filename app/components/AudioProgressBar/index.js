/**
 *
 * AudioProgressBar
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import dynamic from 'next/dynamic';
import getFormattedTimeString from '../../utils/getFormattedTimeString';
import Colors from '../../../theme_config/javascriptColors';
import AudioDramaToggle from '../AudioDramaToggle';

const Slider = dynamic(() => import('rc-slider/lib/Slider'), { ssr: false });

class AudioProgressBar extends React.PureComponent {
	get timeLeft() {
		return getFormattedTimeString(this.props.duration);
	}

	get timePassed() {
		return getFormattedTimeString(this.props.currentTime);
	}

	handleChange = (v) => {
		this.props.setCurrentTime(this.props.duration * (v / 100) || 0);
	};

	render() {
		const percent = 100 * (this.props.currentTime / this.props.duration) || 0;
		return (
			<div
				className={'progress-slider'}
				data-value-dur={this.timeLeft}
				data-value-cur={this.timePassed}
			>
				<Slider
					onChange={this.handleChange}
					handleStyle={{
						border: `2px solid ${Colors.sliderGreen}`,
						backgroundColor: Colors.black,
						top: '5px',
					}}
					railStyle={{
						backgroundColor: Colors.audioProgressBackground,
						height: '4px',
					}}
					trackStyle={{ backgroundColor: Colors.sliderGreen, height: '4px' }}
					value={percent}
					min={0}
					max={100}
				/>
				<AudioDramaToggle />
			</div>
		);
	}
}

AudioProgressBar.propTypes = {
	duration: PropTypes.number,
	currentTime: PropTypes.number,
	setCurrentTime: PropTypes.func,
};

export default AudioProgressBar;
