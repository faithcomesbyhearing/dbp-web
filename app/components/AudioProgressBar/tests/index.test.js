import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import AudioProgressBar from '..';

jest.mock('../../AudioDramaToggle', () => function () {
  return (
<div className={'audio-drama-toggle-container'}>
		<button
			type={'button'}
			id={'drama-button'}
			className={'audio-drama-toggle-button active'}
			onClick={jest.fn()}
		>
			Drama
		</button>
		<button
			type={'button'}
			id={'non-drama-button'}
			className={'audio-drama-toggle-button'}
			onClick={jest.fn()}
		>
			Non-Drama
		</button>
</div>
);
});

const setCurrentTime = jest.fn();
const duration = 300;
const currentTime = 0;

describe('AudioProgressBar Component', () => {
	it('Should match previous snapshot', () => {
		const tree = renderer.create(
			<AudioProgressBar
				duration={duration}
				currentTime={currentTime}
				setCurrentTime={setCurrentTime}
			/>,
		);

		expect(tree).toMatchSnapshot();
	});
	it('Should render a div containing given duration and currentTime', () => {
		const wrapper = mount(
			<AudioProgressBar
				duration={duration}
				currentTime={currentTime}
				setCurrentTime={setCurrentTime}
			/>,
		);
		const style = wrapper.find('.rc-slider-track').get(1).props.style;
		const expectedWidth = `${(100 * (currentTime / duration)).toFixed(0)}%`;

		expect(wrapper.find('.rc-slider-track')).toBeTruthy();
		expect(style).toHaveProperty('width', expectedWidth);
	});
	it('Should render the slider at the correct position', () => {
		const newCurrentTime = 5;
		const wrapper = mount(
			<AudioProgressBar
				duration={duration}
				currentTime={newCurrentTime}
				setCurrentTime={setCurrentTime}
			/>,
		);
		const style = wrapper.find('.rc-slider-track').get(1).props.style;
		const expectedWidth = `${(100 * (newCurrentTime / duration)).toFixed(0)}%`;

		expect(wrapper.find('.rc-slider-track')).toBeTruthy();
		expect(style).toHaveProperty('width', expectedWidth);
	});
});
