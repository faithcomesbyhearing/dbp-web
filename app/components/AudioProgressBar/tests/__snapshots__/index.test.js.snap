// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioProgressBar Component Should match previous snapshot 1`] = `
<div>
  <div
    class="progress-slider"
    data-value-cur="00:00"
    data-value-dur="05:00"
  >
    <div
      class="rc-slider rc-slider-horizontal"
    >
      <div
        class="rc-slider-rail"
        style="background-color: rgb(91, 95, 100); height: 4px;"
      />
      <div
        class="rc-slider-track"
        style="left: 0%; width: 0%; background-color: rgb(98, 177, 130); height: 4px;"
      />
      <div
        class="rc-slider-step"
      />
      <div
        aria-disabled="false"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="rc-slider-handle"
        role="slider"
        style="left: 0%; transform: translateX(-50%); border: 2px solid rgba(98, 177, 130, 1); background-color: rgb(0, 0, 0); top: 5px;"
        tabindex="0"
      />
    </div>
    <div
      class="audio-drama-toggle-container"
    >
      <button
        class="audio-drama-toggle-button active"
        id="drama-button"
        type="button"
      >
        Drama
      </button>
      <button
        class="audio-drama-toggle-button"
        id="non-drama-button"
        type="button"
      >
        Non-Drama
      </button>
    </div>
  </div>
</div>
`;
